server: implement unified status server

dhartunian opened the pull request:
Previously, we had a separate tenant status server that implemented a subset of
status server RPC handlers for SQL tenants to use. This commit modifies the
existing status server to create a single implementation that can be shared
between app and system tenants.

There are two primary reasons why the two implementations diverged:

1. The tenant server simply does not have many capabilities that would allow it
to serve certain StatusServer requests. For example: gossip. This is simple to
reconcile as certain RPCs will return errors. Some additional work is necessary
to ensure that tenant implementations don't panic and return proper responses.
Previously, this was accomplished via the base implementation that would return
"Unimplemented" errors for all handlers, now we will have to modify our
handlers to know when they are run as tenants.

2. Requests that require fan-out to either nodes in a cluster or instances in a
tenant, require different code to execute. This has now been moved behind an
interface called `ServerIterator` that has two implementations: one for nodes,
and another for tenant instances.

Once we have the admin server migrated as well, the full API V2 server can be
implemented on tenants and we should have simpler feature parity between the
two.

Contributes to: https://github.com/cockroachdb/cockroach/issues/80789
Epic: [CRDB-17356](https://cockroachlabs.atlassian.net/browse/CRDB-17356)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91324)
<!-- Reviewable:end -->

knz reviewed (commented):
This design works very nicely. Makes me happy. :100: 
Don't forget to copiously document the various interfaces. 

> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-:-NGI-pUA1lWXLnspIAs9:b-6kvxel)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/auto_upgrade.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGI--bz4VYUhOlPg8ty:-NGI--bz4VYUhOlPg8tz:b32wkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/auto_upgrade.go#L92)):*
> ```Go
> 		log.Errorf(ctx, "failed attempt to upgrade cluster version, error: %v", err)
> 	}
> }
> ```

es
___
*[`pkg/server/auto_upgrade.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGHzwXtAy6uN34YASlN:-NGHzwXtAy6uN34YASlO:b-tbiua7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/auto_upgrade.go#L94)):*
> ```Go
> }
> 
> // nodesStatusWithLiveness is like Nodes but for internal
> ```

I'm friendly to the idea of moving this code, but `auto_upgrade.go` doesn't seem like the right destination. Maybe `node_status.go`?
___
*[`pkg/server/fanout_clients.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGI-Yc7Cf5Q6M1aQp0S:-NGI-Yc7Cf5Q6M1aQp0T:b-2ktgil) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/fanout_clients.go#L22)):*
> ```Go
> )
> 
> type ServerID int32
> ```

why does this and the one below need to be exported?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/auto_upgrade.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGI--bz4VYUhOlPg8ty:-NGI-y5oAOADKf55CKfF:b5rwoij) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/auto_upgrade.go#L92)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

es
</blockquote></details>

sorry mistyped


<!-- Sent from Reviewable.io -->

knz commented:
title: "unify status server between system and secondary tenants"

dhartunian committed:
server: implement unified status server

Previously, we had a separate tenant status server that implemented a subset of
status server RPC handlers for SQL tenants to use. This commit modifies the
existing status server to create a single implementation that can be shared
between app and system tenants.

There are two primary reasons why the two implementations diverged:

1. The tenant server simply does not have many capabilities that would allow it
to serve certain StatusServer requests. For example: gossip. This is simple to
reconcile as certain RPCs will return errors. Some additional work is necessary
to ensure that tenant implementations don't panic and return proper responses.
Previously, this was accomplished via the base implementation that would return
"Unimplemented" errors for all handlers, now we will have to modify our
handlers to know when they are run as tenants.

2. Requests that require fan-out to either nodes in a cluster or instances in a
tenant, require different code to execute. This has now been moved behind an
interface called `ServerIterator` that has two implementations: one for nodes,
and another for tenant instances.

RPCs that only contain system tenant implementations have been migrated to the
systemStatusServer implementation which embeds the base implementation and only
overrides the few handlers that require access to system structs like gossip
and node liveness.

Once we have the admin server migrated as well, the full API V2 server can be
implemented on tenants and we should have simpler feature parity between the
two.

Contributes to: #80789
Resolves: #92615
Epic: CRDB-17356

Release note: None

knz reviewed (commented):
This is pretty elegant. Well done.

> Reviewed 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-:-NHQsSHM-J0kuyEu4Hz6:b-8432sn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/auto_upgrade.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGHzwXtAy6uN34YASlN:-NHQozmQ0oAX3yAUhybv:b-jel8rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/auto_upgrade.go#L94)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I'm friendly to the idea of moving this code, but `auto_upgrade.go` doesn't seem like the right destination. Maybe `node_status.go`?
</blockquote></details>

still relevant
___
*[`pkg/server/fanout_clients.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGI-Yc7Cf5Q6M1aQp0S:-NHQp51pDeIB7BJdEmso:bse2r8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/fanout_clients.go#L22)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

why does this and the one below need to be exported?
</blockquote></details>

probably un-export them.
Otherwise, ensure they are suitably documented.
___
*[`pkg/server/fanout_clients.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQp9R56dtmwoBaAr9o:-NHQp9R56dtmwoBaAr9p:brq43o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/fanout_clients.go#L26)):*
> ```Go
> 
> type ServerIterator interface {
> 	dialNode(
> ```

nit: document each method in the interface for clarity.
___
*[`pkg/server/fanout_clients.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQpvcj3ieUpItsCMxO:-NHQpvcj3ieUpItsCMxP:brfh4cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/fanout_clients.go#L57)):*
> ```Go
> 
> func (t *tenantFanoutClient) getServerIDSQLAddress(id ServerID) (*util.UnresolvedAddr, error) {
> 	return t.getServerIDAddress(id)
> ```

This one gives me pause. We only ever specified that the instance address in the instances table was that of the RPC service.

Since a recent PR, the SQL service can be running on a different port. For the time being, we don't store it in the instances table.
If the SQL service indeed runs on a separate port, it looks like this logic is incorrect.

I realize this _also_ means the current code in `tenant_status.go` is also incorrect:
```go
    // For SQL only servers, the (RPC) Address and SQL address is the same.
    // TODO(#76175): We should split the instance address into SQL and RPC addresses.
    nodeDetails ≔ serverpb.NodeDetails{
      NodeID:     int32(instance.InstanceID),
      Address:    util.MakeUnresolvedAddr("tcp", instance.InstanceAddr),
      SQLAddress: util.MakeUnresolvedAddr("tcp", instance.InstanceAddr),
    }
```

I appreciate that given the existing code was already incorrect, you don't need to fix it here.
However, at the very list your code here should include a reference to that issue #76175.

(It may also be worth scheduling a fix for it sometime soon.)
___
*[`pkg/server/status.go` line 587 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQqNs5ClZlrHYlwIcR:-NHQqNs5ClZlrHYlwIcS:b-dq2mvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/status.go#L587)):*
> ```Go
> 	serverIterator ServerIterator,
> ) *statusServer {
> 	if rpcCtx.TenantID.IsSystem() {
> ```

consider this instead (single code path):
```go
ambient.AddLogTag("status", nil)
ambient.AddLogTag("tenant", rpcCtx.TenantID) 
```
___
*[`pkg/server/status.go` line 2114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQre_2BB9Kv6gOAXR0:-NHQre_3-E0rgcLYTR7H:b7itvw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/status.go#L2114)):*
> ```Go
> 	}
> 
> 	if !s.sqlServer.execCfg.Codec.ForSystemTenant() {
> ```

Can you coordinate with @arulajmani to ensure that this special case (and the code below) gets removed/simplified once #92029 merges. If you don't do it here, please file a followup issue so the MT team / Arul can look at this. Thanks.
___
*[`pkg/server/tenant.go` line 219 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQsJfbEnjNj7VAQDY5:-NHQsJfbEnjNj7VAQDY6:bulu2jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/tenant.go#L219)):*
> ```Go
> 		nil, // should be sAdmin but doesn't "fit" yet
> 		args.db,
> 		nil, // TODO(davidh): what else uses gossip? maybe nothing...
> ```

What do you plan to do about these args within this PR?
___
*[`pkg/server/tenant.go` line 229 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQsGxs9Rp0sDfJAwsM:-NHQsGxs9Rp0sDfJAwsN:b7srkan) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/tenant.go#L229)):*
> ```Go
> 		closedSessionCache,
> 		args.remoteFlowRunner,
> 		args.circularInternalExecutor, // TODO(davidh): does this work??
> ```

i believe that one does, yes.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Still looks good - also some open comments remaining from the last review.

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-:-NI8Sqei2DYp8L66QrQP:buits7l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`-- commits` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NI8LPBsENs0yaTV0BMZ):*
nit: colon not in right place
___
*[`pkg/server/status.go` line 588 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NI8LWPA9aGHFMYAxgeo:-NI8LWPA9aGHFMYAxgep:bblz1pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L588)):*
> ```Go
> 	serverIterator ServerIterator,
> ) *statusServer {
> 	if rpcCtx.TenantID.IsSystem() {
> ```

Is this distinction useful? I don't think we need it any more. You could remove this conditional.

(Ditto in the other place where there's such a thing)


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
This is really great work @dhartunian 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-:-NIDU_1E5cUk_v4Q-W8X:b-25cwzi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/fanout_clients.go` line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDL95LAMTG9xAveRvU:-NIDL95M3uiCn9NBY4wq:bw6v3qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L46)):*
> ```Go
> func (t *tenantFanoutClient) getServerIDAddress(serverID ServerID) (*util.UnresolvedAddr, error) {
> 	id := base.SQLInstanceID(serverID)
> 	instance, err := t.sqlServer.sqlInstanceReader.GetInstance(context.Background(), id)
> ```

nit: can `getServerIDAddress` have a context param, so that we can use it here instead of a background context?



_Code quote:_
```Go
context.Background()
```
___
*[`pkg/server/fanout_clients.go` line 64 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDM0AyEXpxGqepr9IT:-NIDM0AyEXpxGqepr9IU:bxzh98k) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L64)):*
> ```Go
> }
> 
> var _ ServerIterator = &tenantFanoutClient{}
> ```

nit: can we move this to be directly below the `tenantFanoutClient` struct definition? 



_Code quote:_
```Go
var _ ServerIterator = &tenantFanoutClient{}
```
___
*[`pkg/server/fanout_clients.go` line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDNDA85vYaXn0M9QpP:-NIDNDA85vYaXn0M9QpQ:bs4xmyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L92)):*
> ```Go
> 
> func (t *tenantFanoutClient) parseServerID(serverIDParam string) (ServerID, bool, error) {
> 	// No parameter provided or set to local.
> ```

just curious - in what case would we be given an empty string as a `serverIDParam`?



_Code quote:_
```Go
// No parameter provided or set to local.
```
___
*[`pkg/server/fanout_clients.go` line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDPqOE25aeOpv9-Dc2:-NIDPqOE25aeOpv9-Dc3:b-pb9c89) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L177)):*
> ```Go
> 
> func (k kvFanoutClient) parseServerID(serverIDParam string) (ServerID, bool, error) {
> 	n, b, err := parseNodeID(k.gossip, serverIDParam)
> ```

nit: can we rename `b` to `isLocal`? 



_Code quote:_
```Go
b
```
___
*[`pkg/server/fanout_clients.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDNjyX4jf2PmEDSG4j:-NIDNjyY6DnAYwph0R8d:b6trodb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L197)):*
> ```Go
> }
> 
> var _ ServerIterator = &kvFanoutClient{}
> ```

nit: ditto here re: putting it directly below the struct definition.



_Code quote:_
```Go
var _ ServerIterator = &kvFanoutClient{}
```
___
*[`pkg/server/status.go` line 502 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDQSfq4ACfGqXKOUsD:-NIDQSfq4ACfGqXKOUsE:b-wc8iz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L502)):*
> ```Go
> }
> 
> type systemStatusServer struct {
> ```

nit: comments explaining the difference between `statusServer` and `systemStatusServer`



_Code quote:_
```Go
type systemStatusServer struct {
```
___
*[`pkg/server/status.go` line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDS2uA7eZhdI39poxq:-NIDS2uA7eZhdI39poxr:b-ywjf19) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L558)):*
> ```Go
> 			for _, m := range tenantSupportedRPCs {
> 				if m == info.FullMethod {
> 					found = true
> ```

nit:  should we break here?



_Code quote:_
```Go
found = true
```
___
*[`pkg/server/tenant.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDToLB2IfRQ4PlAXpe:-NIDToLB2IfRQ4PlAXpf:b-hs06w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/tenant.go#L203)):*
> ```Go
> 	// the SQL server object until later.
> 
> 	// TODO(davidh): docstring re: above
> ```

nit: a few comment TODOs leftover in this file :upside_down_face: 



_Code quote:_
```Go
	// TODO(davidh): docstring re: above
```


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @knz)

___
*[`-- commits` line 34 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NI8LPBsENs0yaTV0BMZ):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: colon not in right place
</blockquote></details>

done.
___
*[`pkg/server/auto_upgrade.go` line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGHzwXtAy6uN34YASlN:-NIIRMLT3Uo4umya6rxQ:b1ivvvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/auto_upgrade.go#L94)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

still relevant
</blockquote></details>

This method was used in single location in this file so I just inlined the code and removed the need for the custom struct.
___
*[`pkg/server/fanout_clients.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NGI-Yc7Cf5Q6M1aQp0S:-NIISIKGBN2NBp8Spaht:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9721eafa1879bd9225364ebed71cb53175f2a35d/pkg/server/fanout_clients.go#L22)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

probably un-export them.
Otherwise, ensure they are suitably documented.
</blockquote></details>

Done.
___
*[`pkg/server/fanout_clients.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQp9R56dtmwoBaAr9o:-NIITtSA6YxpEHrQ81E0:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/fanout_clients.go#L26)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: document each method in the interface for clarity.
</blockquote></details>

done.
___
*[`pkg/server/fanout_clients.go` line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQpvcj3ieUpItsCMxO:-NIIUUms1P2djOODGm15:b-o6pmm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/fanout_clients.go#L57)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This one gives me pause. We only ever specified that the instance address in the instances table was that of the RPC service.

Since a recent PR, the SQL service can be running on a different port. For the time being, we don't store it in the instances table.
If the SQL service indeed runs on a separate port, it looks like this logic is incorrect.

I realize this _also_ means the current code in `tenant_status.go` is also incorrect:
```go
    // For SQL only servers, the (RPC) Address and SQL address is the same.
    // TODO(#76175): We should split the instance address into SQL and RPC addresses.
    nodeDetails ≔ serverpb.NodeDetails{
      NodeID:     int32(instance.InstanceID),
      Address:    util.MakeUnresolvedAddr("tcp", instance.InstanceAddr),
      SQLAddress: util.MakeUnresolvedAddr("tcp", instance.InstanceAddr),
    }
```

I appreciate that given the existing code was already incorrect, you don't need to fix it here.
However, at the very list your code here should include a reference to that issue #76175.

(It may also be worth scheduling a fix for it sometime soon.)
</blockquote></details>

Thanks for pointing this out. Added TODO.
___
*[`pkg/server/fanout_clients.go` line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDL95LAMTG9xAveRvU:-NIIUIb60d7zv13cI3Bm:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L46)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: can `getServerIDAddress` have a context param, so that we can use it here instead of a background context?
</blockquote></details>

done.
___
*[`pkg/server/fanout_clients.go` line 64 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDM0AyEXpxGqepr9IT:-NIIUfxy-VATmWAr_zws:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L64)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: can we move this to be directly below the `tenantFanoutClient` struct definition? 
</blockquote></details>

Done.
___
*[`pkg/server/fanout_clients.go` line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDNDA85vYaXn0M9QpP:-NIIUmk97DYOAWRpr8yj:b-qak4mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L92)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

just curious - in what case would we be given an empty string as a `serverIDParam`?
</blockquote></details>

The request with nodeID param string comes from the UI so I think it can be omitted and get set to local one automatically.
___
*[`pkg/server/fanout_clients.go` line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDPqOE25aeOpv9-Dc2:-NIIWkh_CNxV4_DkQSyI:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L177)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: can we rename `b` to `isLocal`? 
</blockquote></details>

done.
___
*[`pkg/server/fanout_clients.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDNjyX4jf2PmEDSG4j:-NIIWlRG8lfVCVWzp5ne:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L197)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: ditto here re: putting it directly below the struct definition.
</blockquote></details>

done.
___
*[`pkg/server/status.go` line 587 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQqNs5ClZlrHYlwIcR:-NIIXngK1ElX7Bb-HxFd:b-rnrapz) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/status.go#L587)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

consider this instead (single code path):
```go
ambient.AddLogTag("status", nil)
ambient.AddLogTag("tenant", rpcCtx.TenantID) 
```
</blockquote></details>

done. 
___
*[`pkg/server/status.go` line 2114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQre_2BB9Kv6gOAXR0:-NIIY0qMCyElTLxix2HL:b-u4r6o9) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/status.go#L2114)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Can you coordinate with @arulajmani to ensure that this special case (and the code below) gets removed/simplified once #92029 merges. If you don't do it here, please file a followup issue so the MT team / Arul can look at this. Thanks.
</blockquote></details>

Actually this can be removed because I ended up moving the `TenantRanges` impl on the `systemStatusServer`. It's accessed through the connector interface from the tenant.
___
*[`pkg/server/status.go` line 502 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDQSfq4ACfGqXKOUsD:-NIIX47i3qtwmmm5s2oL:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L502)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: comments explaining the difference between `statusServer` and `systemStatusServer`
</blockquote></details>

done.
___
*[`pkg/server/status.go` line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDS2uA7eZhdI39poxq:-NIIXG-n6lvqUrwLxoM_:b-sw51cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L558)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit:  should we break here?
</blockquote></details>

wowww this code is unused :flushed:. removed. I think it was from an earlier attempt and separating out tenant functionality before I realized I can just rely on struct embedding method resolution.
___
*[`pkg/server/status.go` line 588 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NI8LWPA9aGHFMYAxgeo:-NIIXctz2FG9aafZOFAg:bgjs2qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L588)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Is this distinction useful? I don't think we need it any more. You could remove this conditional.

(Ditto in the other place where there's such a thing)
</blockquote></details>

agree. done.
___
*[`pkg/server/tenant.go` line 219 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NHQsJfbEnjNj7VAQDY5:-NI4LMGOBc2S5r5qYqJP:b-gwcqsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/528d7673974483ed39f359e866f6e52a00a4b085/pkg/server/tenant.go#L219)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

What do you plan to do about these args within this PR?
</blockquote></details>

I realized i can use optional gossip and optional nodeliveness which already exist. will think of something add'l for the rest of the args.
___
*[`pkg/server/tenant.go` line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDToLB2IfRQ4PlAXpe:-NIIZCRZ0wD5KUf_wfpO:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/tenant.go#L203)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: a few comment TODOs leftover in this file :upside_down_face: 
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @dhartunian)

___
*[`pkg/server/fanout_clients.go` line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXe6ct1ff4Oj0Skpqt:-NIXe6ct1ff4Oj0Skpqu:b-r405up) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L41)):*
> ```Go
> // ServerIterator is an interface that provides an abstraction
> // around fanning out requests to different "servers" or "nodes".
> // It is meant to be implemented in two contexts: one for kv
> ```

nit: kv -> KV
___
*[`pkg/server/fanout_clients.go` line 43 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXe5Fd-OhyA_BGCT8J:-NIXe5Fd-OhyA_BGCT8K:bo08fhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L43)):*
> ```Go
> // It is meant to be implemented in two contexts: one for kv
> // servers to fan out to cluster nodes, and another for SQL
> // servers to fan out to sql instances. The implementation of
> ```

nit: sql->SQL
___
*[`pkg/server/fanout_clients.go` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeA957sOY8ubaVxIX:-NIXeA957sOY8ubaVxIY:b-phr7o4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L59)):*
> ```Go
> 	// parseServerID parses the given string as either a node
> 	// or instance ID and returns a bool that is true if the
> 	// id corresponds to the local node or instance that the
> ```

id->ID
___
*[`pkg/server/fanout_clients.go` line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeB2V7KuFN7XNQi70:-NIXeB2V7KuFN7XNQi71:bi2c9jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L62)):*
> ```Go
> 	// call was executed on.
> 	parseServerID(serverIDParam string) (serverID, bool, error)
> 	// getID returns the current node or sql instance ID on the
> ```

sql->SQL
___
*[`pkg/server/fanout_clients.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeCJh-MZBBqpHweex:-NIXeCJh-MZBBqpHweey:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L66)):*
> ```Go
> 	getID() serverID
> 	// getServerIDAddress returns a gRPC address for the given node
> 	// or sql instance.
> ```

ditto
___
*[`pkg/server/fanout_clients.go` line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeD5ID_dbRdV4HjCR:-NIXeD5ID_dbRdV4HjCS:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L69)):*
> ```Go
> 	getServerIDAddress(context.Context, serverID) (*util.UnresolvedAddr, error)
> 	// getServerIDSQLAddress returns a SQL address for the given node
> 	// or sql instance.
> ```

ditto


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @knz)

___
*[`pkg/server/fanout_clients.go` line 41 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXe6ct1ff4Oj0Skpqt:-NIXupKh3sFs6qkjL3fn:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L41)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: kv -> KV
</blockquote></details>

done
___
*[`pkg/server/fanout_clients.go` line 43 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXe5Fd-OhyA_BGCT8J:-NIXupmC9Mv96Ei41j9g:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L43)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: sql->SQL
</blockquote></details>

done
___
*[`pkg/server/fanout_clients.go` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeA957sOY8ubaVxIX:-NIXuqDyCJ9nLxmFMvOq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L59)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

id->ID
</blockquote></details>

done
___
*[`pkg/server/fanout_clients.go` line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeB2V7KuFN7XNQi70:-NIXuthBBgCrLY8Kiexz:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L62)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

sql->SQL
</blockquote></details>

done
___
*[`pkg/server/fanout_clients.go` line 66 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeCJh-MZBBqpHweex:-NIXuuDK5o0Ht4-jVak5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L66)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

done
___
*[`pkg/server/fanout_clients.go` line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIXeD5ID_dbRdV4HjCR:-NIXuuyQ2vyM-p3o_qCl:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f660776de71dc46170e50fb8603ee5e0ac210431/pkg/server/fanout_clients.go#L69)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

abarganier reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Again - great work on this one! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-:-NIZ4o9q8V6aJ3jI0S33:b-erun4m)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian and @knz)

___
*[`pkg/server/fanout_clients.go` line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDNDA85vYaXn0M9QpP:-NIZ4Xw4ESmheJaCsY63:b-6sn7r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/fanout_clients.go#L92)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

The request with nodeID param string comes from the UI so I think it can be omitted and get set to local one automatically.
</blockquote></details>

I see - thanks for explaining!
___
*[`pkg/server/status.go` line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91324#-NIDS2uA7eZhdI39poxq:-NIZ4iPY-COBgAbkqcWx:b5k9dkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e9ad16ae70b1c26fa96adbae83697cd1360be86/pkg/server/status.go#L558)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

wowww this code is unused :flushed:. removed. I think it was from an earlier attempt and separating out tenant functionality before I realized I can just rely on struct embedding method resolution.
</blockquote></details>

Hah fair enough! Thanks for cleaning it up.


<!-- Sent from Reviewable.io -->

dhartunian commented:
TFTRs!
bors r=knz,abarganier
